Лабораторная работа 3

RuslanGorelenkov opened the pull request.

RuslanGorelenkov committed:
Лабораторная работа 3

frek17 commented on Лабораторная работа 3/task6.py:
```diff
@@ -0,0 +1,10 @@
+    if number > max_number:
+        max_number = number
+        count = index
+list_numbers[index], list_numbers[9] = list_numbers[9], list_numbers[index]
```
Откуда 9?

frek17 replied:
Зачем переменная `count`?

RuslanGorelenkov replied:
9 это индекс 10 элемента. Count переменная, которая отвечает за индекс максимального элемента.

frek17 replied:
> 9 это индекс 10 элемента

Так к чему хардкодинг? Откуда взялось число 9?

frek17 replied:
> Count переменная, которая отвечает за индекс максимального элемента

Название переменной должно отражать цель переменной. `count` переводится как считать, хотя смысл переменной - индекс максимального числа/элемента

RuslanGorelenkov replied:
В задании сказано, что дан целочисленный список из 10 элементов. Соответственно, последний элемент будет под номером 9. Программа сработала верно. Count мне действительно следовало заменить, например на max_index.

frek17 replied:
Откуда 10 элементов? Их ведь не 10

RuslanGorelenkov replied:
в задании сказано 10. Если их не 10 в условии, но программа работает, это не моя ошибка

frek17 replied:
Вы не поняли. Зачем вам переменные max_number и count, если вы используете 9? В чем логика?

RuslanGorelenkov replied:
мне нужно поменять максимальный элемент и последний. Max_number и count нужны для его поиска. А 9 нужно для того, чтобы собственно поменять последний и максимальный.

frek17 replied:
Так если вы max_number и count вообще не используете после цикла, зачем вы их находили?

RuslanGorelenkov replied:
Они нужны для цикла

frek17 replied:
Что значит "для цикла"? Переменные и в принципе код пишутся, чтобы получить какой-то результат, а для этого у каждого действия должен быть смысл. Так какой смысл у этих двух переменных, зачем мы нашли их в цикле? И вообще что в них содержится в итоге?

RuslanGorelenkov replied:
я переделал программу, в этой была ошибка

frek17 replied:
Вижу, теперь всё хорошо

frek17 commented on Лабораторная работа 3/task8.py:
```diff
@@ -0,0 +1,15 @@
+while money_capital >= 0:
+    month += 1
+    spend *= 1 + increase
+    delta += salary - spend
```
Почему `+=`?

RuslanGorelenkov replied:
+= мы заменяем выражение на комбинированный оператор присваивания
Например :
month = month + 1 - > month += 1

frek17 replied:
Спасибо за ответ, но вопрос больше про саму дельту. Зачем в принципе каждый ход прибавлять дельту к самой себе? А потом эту увеличенную дельту прибавлять (точнее вычитать) к `money_capital`?

RuslanGorelenkov replied:
Да, можно было обойтись и без переменной delta и прибавлять сразу к money_capital. Я ввел ее для того, чтобы мне было удобней.

frek17 replied:
И снова не поняли. В чем смысл дельты и зачем к ней прибавлять это значение?

RuslanGorelenkov replied:
delta это то, сколько денег теряется каждый месяц

frek17 replied:
Ну так почему дельта меняется таким образом? Логики в этом выражении нет. Жду объяснение

RuslanGorelenkov replied:
а каким образом она должна меняться? В первый месяц мы уходим в минус на delta, за второй и первый на delta=delta+salary-spend. В чем проблема?

frek17 replied:
Хорошо, объясню вам ваш же код. Смысл дельты — разница трат и дохода, то есть на сколько будет уменьшаться финансовая подушка каждый месяц. Тогда для каждого конкретного месяца дельта это: траты (с учетом инфляции) минус зарплата. У вас же для последующего месяца берется эта разница, да, но еще предыдущее значение дельты, ведь у вас:
delta = delta + salary - spend
Видите, вы не перезаписываете значение дельты для нового месяца, а вы тянете еще дельту с прошлого месяца? Понятна ошибка?

RuslanGorelenkov replied:
Понял, переделаю. Спасибо за долгожданную помощь.

frek17 replied:
А теперь еще момент: мы два раза делаем одно и то же — задаем дельту одной формулой перед циклом (для первого месяца) и в цикле. Логичнее задавать дельту только в самом начале цикла. Понятно, что я говорю?

RuslanGorelenkov replied:
Да, я пробовал так делать, но тогда получается ответ 6, цикл while с предусловием посчитает 6 месяц, т.к. money_capital в начале 6го месяца будет положительным, но не достаточным чтобы прожить 6ой месяц. Тогда нужно использовать цикл с постусловием. Например вот так
![image](https://user-images.githubusercontent.com/115419110/198881310-b1ba52dd-e2a8-4e70-8020-1116e489a668.png)

RuslanGorelenkov replied:
Могу переделать программу с другим циклом, если скриншот не подходит.

frek17 replied:
Вот код на скриншоте выдает ответ 5, это правильно

frek17 replied:
В общем, решение принимаю, баллы сейчас проставлю

RuslanGorelenkov replied:
Вот такое выдает ответ 6
![image](https://user-images.githubusercontent.com/115419110/199087281-458eccd2-9542-43de-843d-bd72eb20679d.png)

frek17 replied:
Да, смотрите, просто в начале 6-го месяца ещё какие-то деньги есть. И здесь получается мы посчитаем этот месяц, хотя в нем выйдем в минус

frek17 replied:
Проще вот как выше было, бесконечный цикл и break, по логике всё ясно

frek17 commented on Лабораторная работа 3/task9.py:
```diff
@@ -0,0 +1,12 @@
+
+need_money = 0  # количество денег, чтобы прожить 10 месяцев
+
+for month in range(1, months + 1):
```
Если написать `for month in range(months):` что-нибудь изменится?

RuslanGorelenkov replied:
Ничего не изменится
Количество шагов в цикле for останется прежним (10). Range(months) создаёт последовательность элементов от 0 до 9. Range(1, months +1) создаёт последовательность с элементами от 1 до 10

frek17 replied:
Если же запись по смыслу такая же (учитывая то, что сам счётчик мы никак не используем), лучше использовать короткую запись. В целом ничего страшного, конечно, но важно лишнее не писать

frek17 commented:
Ping

RuslanGorelenkov commented:
> пинг

?

frek17 commented:
На вопросы ответьте, пожалуйста

RuslanGorelenkov commented:
> На вопросы ответьте, пожалуйста

мне не задали вопросы

frek17 commented:
Вы разве не видите комментарии выше?
Прикрепите скриншот этой страницы

RuslanGorelenkov commented:
> Вы разве не видите комментарии выше? Прикрепите скриншот этой страницы

![Screenshot_20221026-162350_Chrome](https://user-images.githubusercontent.com/115419110/198037538-b60d1983-8410-4d05-bd7e-d147d6761200.jpg)

frek17 commented:
Попробуйте в Pull Request перейти

RuslanGorelenkov commented:
у меня нигде нет вопросов

RuslanGorelenkov commented:
![Screenshot_20221026-162828_Chrome](https://user-images.githubusercontent.com/115419110/198039144-7dba8d7a-ecec-4157-bf83-6b83b65b7a02.jpg)

RuslanGorelenkov commented:
Почему мне уже выставили оценку за лабораторную?

RuslanGorelenkov commented:
> Попробовать в Pull Request перейти

Как мне повысить мою оценку за лабораторную?

RuslanGorelenkov commented:
Я ответил

RuslanGorelenkov commented:
Ответил

RuslanGorelenkov commented:
Я ответил

RuslanGorelenkov commented:
Почему у меня стоит 2 из 4 баллов?

frek17 commented:
Что-то было с гитом и не опубликовались комментарии — моя вина. 2 из 4 из-за того, что пока 8 и 9 я не принимаю. Нужно четко ответить на все вопросы и после поправить код, после я уже зачту 8 и 9 задания. Насчет поздней сдачи пока не переживайте. Пинговать каждый день не стоит, я не каждый день сюда захожу

RuslanGorelenkov commented:
Зачем мне нужно поправлять код, если у меня правильные ответы?

frek17 commented:
Во-первых, "поправлять". Во-вторых, давайте люди с таким же подходом будут лечить людей, строить мосты, конструировать самолеты. И будет всё совершенно прекрасно, ведь важен "ответ", а не "логичное решение". Что думаете?
По заданиям я напишу в соответствующих местах.

RuslanGorelenkov commented:
> Во-первых, "поправлять". Во-вторых, давайте люди с таким же подходом будут лечить людей, строить мосты, конструировать самолеты. И будет всё совершенно прекрасно, ведь важен "ответ", а не "логичное решение". Что думаете? По заданиям я напишу в соответствующих местах.

Давайте вы будете проверять орфографию на другом курсе, а если мой самолёт летает, а мост держится, то наверно проблем нет. Что думаете?

RuslanGorelenkov committed:
Update task8.py

 Я изменил delta в цикле, теперь каждый раз она не суммируется с прошлой. Ответ получился 5

RuslanGorelenkov committed:
Update task6.py

Я поменял count на max_index и сделал для нумерации с конца списка.

RuslanGorelenkov commented:
Я переделал код для 6 и 8 заданий, так думаю будет быстрее и проще для нас обоих. Они находятся в commits
